File: src/main/java/macro303/keschet/players/console/Console.java
package macro303.keschet.players.console;

import macro303.keschet.Colour;
import macro303.keschet.Coordinates;
import macro303.keschet.Util;
import macro303.keschet.board.Board;
import macro303.keschet.board.Square;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Created by Macro303 on 2018-02-12.
 */
abstract class Console {
	private static final Logger LOGGER = LogManager.getLogger(Console.class);

	static void showTitle(@NotNull String title, @NotNull Colour colour) {
		String string = IntStream.range(0, title.length() + 4).mapToObj(i -> "=").collect(Collectors.joining());
		colourConsole(string, colour);
		colourConsole("  " + title + "  ", colour);
		colourConsole(string, colour);
	}

	private static void colourConsole(@NotNull String message, @NotNull Colour messageColour) {
		System.out.print(messageColour.getColourCode());
		System.out.print(message);
		System.out.println(Colour.RESET.getColourCode());
	}

	private static void colourConsole(@NotNull String title, @NotNull Colour titleColour, @NotNull String message, @NotNull Colour messageColour) {
		System.out.print(titleColour.getColourCode());
		System.out.print(title);
		colourConsole(message, messageColour);
	}

	static void showMessage(@NotNull String message) {
		colourConsole(message, Colour.WHITE);
		LOGGER.info(message);
	}

	static void showError(@NotNull String message) {
		colourConsole(message, Colour.MAGENTA);
		LOGGER.warn(message);
	}

	static void drawBoard(@NotNull Board board, boolean colourSides) {
		for (int row = -1; row < Util.SIZE; row++) {
			for (int col = -1; col < Util.SIZE; col++) {
				System.out.print(Colour.GREEN.getColourCode());
				if (row == -1 && col == -1)
					System.out.print("  ");
				else if (col == -1)
					System.out.print(row + " ");
				else if (row == -1)
					System.out.print(" " + col + " ");
				else {
					System.out.print(Colour.YELLOW.getColourCode());
					if (colourSides) {
						if (row < 3)
							System.out.print(Colour.BLUE.getColourCode());
						else if (row > 6)
							System.out.print(Colour.RED.getColourCode());
					}
					Square square = board.getSquare(new Coordinates(row, col));
					assert square != null;
					if (square.getPiece() == null)
						System.out.print(" ~ ");
					else
						System.out.print(" " + square.getPiece().getTeamColour().getColourCode() + square.getPiece().getSymbol() + " ");
				}
			}
			System.out.println();
		}
		System.out.print(Colour.RESET.getColourCode());
	}
}

File: Base/src/main/kotlin/macro303/base/Console.kt
package macro303.base

import macro303.base.pieces.*
import org.apache.logging.log4j.LogManager
import java.util.*

/**
 * Created by Macro303 on 2017-10-20.
 */
object Console {
	private val LOGGER = LogManager.getLogger(Console::class.java)

	private val headingColour = Colour.MAGENTA
	private val importantColour = Colour.YELLOW
	private val messageColour = Colour.WHITE
	private val squareColour = Colour.GREEN

	private fun colourConsole(
		title: String? = null,
		titleColour: Colour = headingColour,
		message: String?,
		messageColour: Colour
	) {
		if (title != null) {
			print(titleColour.colourCode)
			print(title)
		}
		print(messageColour.colourCode)
		print(message)
		println(Colour.RESET.colourCode)
	}

	fun showTitle(title: String?, colour: Colour = headingColour) {
		colourConsole(message = "=".repeat(title?.length?.plus(4) ?: 8), messageColour = colour)
		colourConsole(message = "  $title  ", messageColour = colour)
		colourConsole(message = "=".repeat(title?.length?.plus(4) ?: 8), messageColour = colour)
	}

	fun showMessage(message: String?) = colourConsole(message = message, messageColour = messageColour)

	fun showError(message: String?) = colourConsole(message = message, messageColour = importantColour)

	fun showValue(title: String?, message: String) = colourConsole(
		title = "$title: ",
		titleColour = importantColour,
		message = message,
		messageColour = messageColour
	)

	fun showSquare(piece: Piece?) {
		print((piece?.teamColour ?: squareColour).colourCode)
		if (piece != null)
			print(" ${getSymbol(piece = piece)} ")
		else
			print(" ~ ")
		print(Colour.RESET.colourCode)
	}

	fun showRules() {
		showTitle(title = "Rules")
		showMessage(message = "At the start of the game a player may place any piece in any square in the first 3 rows at the player's end of the board.")
		showMessage(message = "A piece is taken if the square it occupies is occupied by an opposing piece.")
		showPieces()
	}

	fun helpMenu(input: String) {
		when {
			input.equals("Help", ignoreCase = true) -> showHelp()
			input.contains("pieces", ignoreCase = true) -> showPieces()
			input.contains("rules", ignoreCase = true) -> showRules()
			else -> {
				val temp: Piece? = when {
					input.contains(
						"<${getSymbol(Archer::class.java)}>",
						ignoreCase = true
					) || input.contains(Archer::class.java.simpleName, ignoreCase = true) -> {
						Archer(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Emperor::class.java)}>",
						ignoreCase = true
					) || input.contains(Emperor::class.java.simpleName, ignoreCase = true) -> {
						Emperor(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(General::class.java)}>",
						ignoreCase = true
					) || input.contains(General::class.java.simpleName, ignoreCase = true) -> {
						General(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Lancer::class.java)}>",
						ignoreCase = true
					) || input.contains(Lancer::class.java.simpleName, ignoreCase = true) -> {
						Lancer(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Merchant::class.java)}>",
						ignoreCase = true
					) || input.contains(Merchant::class.java.simpleName, ignoreCase = true) -> {
						Merchant(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Scholar::class.java)}>",
						ignoreCase = true
					) || input.contains(Scholar::class.java.simpleName, ignoreCase = true) -> {
						Scholar(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Spearman::class.java)}>",
						ignoreCase = true
					) || input.contains(Spearman::class.java.simpleName, ignoreCase = true) -> {
						Spearman(teamColour = Colour.RESET)
					}
					input.contains(
						"<${getSymbol(Thief::class.java)}>",
						ignoreCase = true
					) || input.contains(Thief::class.java.simpleName, ignoreCase = true) -> {
						Thief(teamColour = Colour.RESET)
					}
					else -> null
				}
				if (temp != null)
					showPiece(temp)
			}
		}
	}

	private fun showHelp() {
		showTitle(title = "Help")
		showValue(title = "Help Rules", message = "Shows the rules.")
		showValue(title = "Help Pieces", message = "Shows you all the pieces.")
		showValue(
			title = "Help <Symbol>",
			message = "Shows you all the information about the piece with that symbol (Symbol must be inside <>)."
		)
		showValue(title = "Help Name", message = "Shows you all the information about the piece with that name.")
	}

	private fun showPieces() {
		showPiece(piece = Archer(teamColour = Colour.RESET))
		showPiece(piece = Emperor(teamColour = Colour.RESET))
		showPiece(piece = General(teamColour = Colour.RESET))
		showPiece(piece = Lancer(teamColour = Colour.RESET))
		showPiece(piece = Merchant(teamColour = Colour.RESET))
		showPiece(piece = Scholar(teamColour = Colour.RESET))
		showPiece(piece = Spearman(teamColour = Colour.RESET))
		showPiece(piece = Thief(teamColour = Colour.RESET))
	}

	private fun showPiece(piece: Piece) {
		showTitle(title = "${piece::class.java.simpleName} Help")
		showValue(title = "Piece", message = piece::class.java.simpleName)
		showValue(title = "Symbol", message = getSymbol(piece = piece))
		showValue(title = "Max Distance", message = piece.maxDistance.toString())
		showValue(title = "Valid Directions", message = Arrays.toString(piece.validDirections))
		when (piece) {
			is Emperor -> showValue(
				title = "Ability",
				message = "The game is won if the Emperor is taken or if the Emperor is the only piece remaining to the losing player"
			)
			is Merchant -> showValue(
				title = "Ability",
				message = "The Merchant can move to any vacant square surrounding the square occupied by the Emperor, if the route is unobstructed by another piece."
			)
			is Scholar -> showValue(
				title = "Ability",
				message = "Any piece in an surrounding square to the Scholar is protected and cannot be taken."
			)
			is Thief -> showValue(
				title = "Ability",
				message = "Any piece taken by the Thief is then placed back on the board under the player's control in one of the surrounding squares."
			)
		}
	}

	private fun getSymbol(piece: Piece): String {
		return getSymbol(clazz = piece::class.java)
	}

	private fun getSymbol(clazz: Class<*>): String {
		return clazz.simpleName.subSequence(startIndex = 0, endIndex = 2).toString()
	}
}
